File: cdk/src/Cdk/EBStack.cs
using Amazon.CDK;
using Amazon.CDK.AWS.ElasticBeanstalk;
using Amazon.CDK.AWS.IAM;
using Amazon.CDK.AWS.S3.Assets;
using Constructs;
using System.Collections.Generic;

namespace Cdk
{
    public class EBEnvProps : StackProps
    {
        public string MinSize { get; set; } = "1";
        public string MaxSize { get; set; } = "1";
        public string InstanceTypes { get; set; } = "t2.micro";
        public string SolutionStackName { get; set; } = "64bit Amazon Linux 2023 v3.2.2 running .NET 8";
        public required string AppName { get; set; }
        public required string EnvName { get; set; }
        public required string Path { get; set; }
        public List<CfnEnvironment.OptionSettingProperty> EnvironmentProperties { get; set; } = new List<CfnEnvironment.OptionSettingProperty>();
    }

    public class EBStack : Stack
    {
        public EBStack(Construct scope, string id, EBEnvProps props) : base(scope, id, props)
        {
            // Create an S3 asset from the source directory
            var webAppZipArchive = new Asset(this, "WebAppZip", new AssetProps
            {
                Path = props.Path
            });

            // Create an Elastic Beanstalk application
            var app = new CfnApplication(this, "Application", new CfnApplicationProps
            {
                ApplicationName = props.AppName
            });

            // Create an application version linked to the S3 asset
            var appVersionProps = new CfnApplicationVersion(this, "AppVersion", new CfnApplicationVersionProps
            {
                ApplicationName = props.AppName,
                SourceBundle = new CfnApplicationVersion.SourceBundleProperty
                {
                    S3Bucket = webAppZipArchive.S3BucketName,
                    S3Key = webAppZipArchive.S3ObjectKey
                }
            });

            // Ensure the application exists before the application version is created
            appVersionProps.AddDependency(app);

            // Create an IAM role for Elastic Beanstalk instances
            var myRole = new Role(this, $"{props.AppName}-aws-elasticbeanstalk-ec2-role", new RoleProps
            {
                AssumedBy = new ServicePrincipal("ec2.amazonaws.com")
            });

            myRole.AddManagedPolicy(ManagedPolicy.FromAwsManagedPolicyName("AWSElasticBeanstalkWebTier"));

            // Create an instance profile for the Elastic Beanstalk environment
            var myProfileName = $"{props.AppName}-InstanceProfile";
            var instanceProfile = new CfnInstanceProfile(this, myProfileName, new CfnInstanceProfileProps
            {
                InstanceProfileName = myProfileName,
                Roles = [myRole.RoleName]
            });

            // Configure Elastic Beanstalk environment options
            var optionSettingProperties = new List<CfnEnvironment.OptionSettingProperty>
            {
                new CfnEnvironment.OptionSettingProperty
                {
                    Namespace = "aws:autoscaling:launchconfiguration",
                    OptionName = "IamInstanceProfile",
                    Value = myProfileName
                },
                new CfnEnvironment.OptionSettingProperty
                {
                    Namespace = "aws:autoscaling:asg",
                    OptionName = "MinSize",
                    Value = props.MinSize
                },
                new CfnEnvironment.OptionSettingProperty
                {
                    Namespace = "aws:autoscaling:asg",
                    OptionName = "MaxSize",
                    Value = props.MaxSize
                },
                new CfnEnvironment.OptionSettingProperty
                {
                    Namespace = "aws:ec2:instances",
                    OptionName = "InstanceTypes",
                    Value = props.InstanceTypes
                }
            };

            foreach (var envProp in props.EnvironmentProperties)
            {
                optionSettingProperties.Add(new CfnEnvironment.OptionSettingProperty
                {
                    Namespace = "aws:elasticbeanstalk:application:environment",
                    OptionName = envProp.OptionName,
                    Value = envProp.Value
                });
            }

            // Create an Elastic Beanstalk environment
            var elbEnv = new CfnEnvironment(this, "Environment", new CfnEnvironmentProps
            {
                EnvironmentName = props.EnvName,
                ApplicationName = app.ApplicationName,
                SolutionStackName = props.SolutionStackName,
                OptionSettings = optionSettingProperties.ToArray(),
                VersionLabel = appVersionProps.Ref,
            });
        }
    }
}

File: cdk-ts/lib/eb-stack.ts
import * as cdk from 'aws-cdk-lib';
import { CfnApplication, CfnApplicationVersion, CfnEnvironment } from 'aws-cdk-lib/aws-elasticbeanstalk';
import { CfnInstanceProfile, ManagedPolicy, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { Asset } from 'aws-cdk-lib/aws-s3-assets';
import { Construct } from 'constructs';

export interface EBEnvProps extends cdk.StackProps {
  minSize?: string;
  maxSize?: string;
  instanceTypes?: string;
  solutionStackName?: string;
  appName: string;
  envName: string;
  path: string;
  environmentProperties?: { optionName: string; value: string }[];
}

export class EBStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: EBEnvProps) {
    super(scope, id, props);

    // Default property values
    const minSize = props.minSize ?? '1';
    const maxSize = props.maxSize ?? '1';
    const instanceTypes = props.instanceTypes ?? 't2.micro';
    const solutionStackName = props.solutionStackName ?? '64bit Amazon Linux 2023 v3.2.2 running .NET 8';

    // Create an S3 asset from the source directory
    const webAppZipArchive = new Asset(this, 'WebAppZip', {
      path: props.path,
    });

    // Create an Elastic Beanstalk application
    const app = new CfnApplication(this, 'Application', {
      applicationName: props.appName,
    });

    // Create an application version linked to the S3 asset
    const appVersionProps = new CfnApplicationVersion(this, 'AppVersion', {
      applicationName: props.appName,
      sourceBundle: {
        s3Bucket: webAppZipArchive.s3BucketName,
        s3Key: webAppZipArchive.s3ObjectKey,
      },
    });

    // Ensure the application exists before the application version is created
    appVersionProps.addDependency(app);

    // Create an IAM role for Elastic Beanstalk instances
    const myRole = new Role(this, `${props.appName}-aws-elasticbeanstalk-ec2-role`, {
      assumedBy: new ServicePrincipal('ec2.amazonaws.com'),
    });

    myRole.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('AWSElasticBeanstalkWebTier'));

    // Create an instance profile for the Elastic Beanstalk environment
    const myProfileName = `${props.appName}-InstanceProfile`;
    const instanceProfile = new CfnInstanceProfile(this, myProfileName, {
      instanceProfileName: myProfileName,
      roles: [myRole.roleName],
    });

    // Configure Elastic Beanstalk environment options
    const optionSettingProperties: CfnEnvironment.OptionSettingProperty[] = [
      {
        namespace: 'aws:autoscaling:launchconfiguration',
        optionName: 'IamInstanceProfile',
        value: myProfileName,
      },
      {
        namespace: 'aws:autoscaling:asg',
        optionName: 'MinSize',
        value: minSize,
      },
      {
        namespace: 'aws:autoscaling:asg',
        optionName: 'MaxSize',
        value: maxSize,
      },
      {
        namespace: 'aws:ec2:instances',
        optionName: 'InstanceTypes',
        value: instanceTypes,
      },
    ];

    // Add custom environment properties
    if (props.environmentProperties) {
      props.environmentProperties.forEach((envProp) => {
        optionSettingProperties.push({
          namespace: 'aws:elasticbeanstalk:application:environment',
          optionName: envProp.optionName,
          value: envProp.value,
        });
      });
    }

    // Create an Elastic Beanstalk environment
    new CfnEnvironment(this, 'Environment', {
      environmentName: props.envName,
      applicationName: app.applicationName!,
      solutionStackName: solutionStackName,
      optionSettings: optionSettingProperties,
      versionLabel: appVersionProps.ref,
    });
  }
}
